client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,348 @@
+          httpclient::GeneralClientConnection::factory(
+              client.getCommFeaturePhase(), newEndpoint, 30, 60, 3,
+              sslProtocol));
+
+      if (connection == nullptr) {
+        continue;
+      }
+      httpclient::SimpleHttpClientParams const params(30, false);
+      _httpClient = std::make_unique<httpclient::SimpleHttpClient>(
+          connection.get(), params);
+    }
+    if (_httpClient != nullptr) {
+      if (!isOriginalUrl) {
+        auto& clientParams = _httpClient->params();
+        clientParams.setRequestTimeout(30);
+      }
+      lk.unlock();
+      if (isOriginalUrl) {
+        std::unique_ptr<httpclient::SimpleHttpResult> response(
+            _httpClient->request(rest::RequestType::POST, relativeUrl,
+                                 compressedBody.data(), compressedBody.size(),
+                                 headers));
+      }
+      std::unique_ptr<httpclient::SimpleHttpResult> response(
+          _httpClient->request(rest::RequestType::POST, relativeUrl,
+                               body.data(), body.size(), headers));
+      lk.lock();
+      result = this->checkHttpResponse(response);
+      if (result.ok()) {
+        auto returnCode = response->getHttpReturnCode();
+        if (returnCode == 200) {
+          if (!isOriginalUrl) {
+            testResponseBuilder.add(response->getBodyVelocyPack()->slice());
+          }
+          break;
+        } else if (returnCode == 301 || returnCode == 302 ||
+                   returnCode == 307) {
+          if (numRedirects < maxRedirects) {
+            bool found;
+            url = response->getHeaderField(StaticStrings::Location, found);
+            if (found) {
+              numRedirects++;
+            } else {
+              url = reqUrl;
+              numRedirects = 0;
+            }
+          } else {
+            url = reqUrl;
+            numRedirects = 0;
+          }
+        }
+      } else if (result.errorNumber() == TRI_ERROR_INTERNAL) {
+        numRedirects = 0;
+      } else {
+        // do not retry if 401, 403, 404, etc.
+        break;
+      }
+      if (numRedirects == 0) {
+        timeoutInSecs *= 3;
+        timeoutInSecs = std::min<uint32_t>(600, timeoutInSecs);
+        _runCondition.wait_for(lk, std::chrono::seconds(timeoutInSecs),
+                               [this] { return _server.isStopping(); });
+      }
+      if (isOriginalUrl) {
+        _httpClient.reset(nullptr);
+      }
+    }
+  }
+  if (!testResponseBuilder.isEmpty()) {
+    return testResponseBuilder;
+  } else {
+    return std::nullopt;
+  }
+}
+
+void TelemetricsHandler::getTelemetricsInfo(VPackBuilder& builder) {
+  // as this is for printing the telemetrics result in the tests, we send the
+  // telemetrics object if the request returned ok and the error if not to
+  // parse the error in the tests
+  std::unique_lock lk(_mtx);
+
+  if (_telemetricsFetchResponse.ok() && !_telemetricsFetchedInfo.isEmpty()) {
```

Comment:
Can there be a race here, e.g. if no telemetrics data have been fetched **yet**, but this function is already called?